homeassistant/components/smartthings/config_flow.py

```diff
@@ -1,298 +1,89 @@
-        """Create a new instance of the flow handler."""
-        self.access_token: str | None = None
-        self.oauth_client_secret = None
-        self.oauth_client_id = None
-        self.installed_app_id = None
-        self.refresh_token = None
-        self.endpoints_initialized = False
-
-    async def async_step_import(self, import_data: None) -> ConfigFlowResult:
-        """Occurs when a previously entry setup fails and is re-initiated."""
-        return await self.async_step_user(import_data)
-
-    async def async_step_user(
-        self, user_input: dict[str, Any] | None = None
-    ) -> ConfigFlowResult:
-        """Validate and confirm webhook setup."""
-        if not self.endpoints_initialized:
-            self.endpoints_initialized = True
-            await setup_smartapp_endpoint(
-                self.hass, len(self._async_current_entries()) == 0
+    VERSION = 3
+    DOMAIN = DOMAIN
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
+
+    @property
+    def extra_authorize_data(self) -> dict[str, Any]:
+        """Extra data that needs to be appended to the authorize url."""
+        return {"scope": " ".join(SCOPES)}
+
+    async def async_oauth_create_entry(self, data: dict[str, Any]) -> ConfigFlowResult:
+        """Create an entry for SmartThings."""
+        client = SmartThings(session=async_get_clientsession(self.hass))
+        client.authenticate(data[CONF_TOKEN][CONF_ACCESS_TOKEN])
+        locations = await client.get_locations()
+        location = locations[0]
+        # We pick to use the location id as unique id rather than the installed app id
+        # as the installed app id could change with the right settings in the SmartApp
+        # or the app used to sign in changed for any reason.
+        await self.async_set_unique_id(location.location_id)
+        if self.source != SOURCE_REAUTH:
+            self._abort_if_unique_id_configured()
+
+            return self.async_create_entry(
+                title=location.name,
+                data={**data, CONF_LOCATION_ID: location.location_id},
             )
-        webhook_url = get_webhook_url(self.hass)
 
-        # Abort if the webhook is invalid
-        if not validate_webhook_requirements(self.hass):
-            return self.async_abort(
-                reason="invalid_webhook_url",
-                description_placeholders={
-                    "webhook_url": webhook_url,
-                    "component_url": (
-                        "https://www.home-assistant.io/integrations/smartthings/"
-                    ),
+        if (
+            (entry := self._get_reauth_entry())
+            and CONF_TOKEN not in entry.data
+            and entry.data[CONF_LOCATION_ID] != location.location_id
+        ):
+            return self.async_abort(reason="reauth_location_mismatch")
+
+        if (entry := self._get_reauth_entry()) and CONF_TOKEN not in entry.data:
```

Comment:
duplicated code `(entry := self._get_reauth_entry()) and CONF_TOKEN not in entry.data` from line 54

Also, by reloading the config entry without a token, it will raise ConfigEntryAuthFailed... So instead of reloading can we offer the reauth flow directly?

Reply:
This is already during reauth, so `data` has the `CONF_TOKEN`, and we update in the `async_update_reload_and_abort`